### PR TITLE
Meta+Userland: Compile GML to C++ for ~~fun and profit~~ correctness and ergonomity

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
**TL;DR: A lot more static validation of GML code**

This is all achieved by generating C++ code from GML which creates the widget structure. The code lives as an implementation of a constructor function on the top-level widget, meaning that besides the CMake setup it only requires the user to declare this function and a public constructor in order to work. For example, Window::set_main_widget will immediately pick up the new constructor and use it.

Some other advantages besides correctness are the reduced distribution of special cases (such as TabWidgets and ScrollableContainerWidgets) and error propagation & new string by default.

While I have opted to keep the old dynamic implementation as it is needed for GMLPlayground, I have removed it from the documentation to encourage new users to use the static implementation.

I recognize that there is some danger of diverging implementations. Because most apps already use static GML, dynamic GML can be considered a bonus feature and the properties of static GML are now normative. (It may have bugs of course, but whenever there is a complaint about "my gml works in playground but not in my app!", we should make sure the playground follows static behavior, even if both behaviors are valid.)

### CODE REVIEW

File: Meta/CMake/code_generators.cmake [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # Functions for generating sources using host tools
 #
 
-function(compile_gml source output string_name)
+function(stringify_gml source output string_name)
```
Comment (linusg) [inline Meta/CMake/code_generators.cmake:5]: Let's do this upfront please so we don't have to introduce "new_compile_gml" temporarily

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include <LibMain/Main.h>
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    if (arguments.argc != 2) {
+        outln("usage: {} <GML file>", arguments.strings[0]);
+        return 1;
+    }
+
+    auto gml_file_name = arguments.strings[1];
```
Comment (linusg) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:22]: Let's not get into the habit of doing direct `arguments` access, use an ArgsParser for nice help & stuff.

File: Userland/Libraries/LibGUI/GML/Parser.cpp [hunk 3]
```diff
@@ -131,4 +134,201 @@ ErrorOr<NonnullRefPtr<GMLFile>> parse_gml(StringView string)
     return file;
 }
 
+enum class UseObjectConstructor : bool {
```
Comment (linusg) [inline Userland/Libraries/LibGUI/GML/Parser.cpp:137]: All of this hardly belongs into Parser.cpp :)

File: Userland/Libraries/LibGUI/GML/Parser.h [hunk 4]
```diff
@@ -14,4 +14,7 @@ namespace GUI::GML {
 
 ErrorOr<NonnullRefPtr<GMLFile>> parse_gml(StringView);
 
+// Generates a C++ file that contains a function which creates the specified widget structure and returns it.
+ErrorOr<String> generate_cpp_loader(NonnullRefPtr<GMLFile> gml);
```
Comment (linusg) [inline Userland/Libraries/LibGUI/GML/Parser.h:18]: The "loader" part of the name is confusing, just "generate_cpp" seems fine

File: Userland/Libraries/LibGUI/GML/Parser.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Libraries/LibGUI/GML/Parser.cpp:137]: Since it's not generally useful for anything else depending on LibGUI, I'd put this into GMLCompiler directly.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:22]: That's another FIXME for the other code generators

Comment (linusg) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:22]: https://github.com/SerenityOS/serenity/blob/4d68dcfca0487b7cc0c421932f4d4a1c8c44ee5b/Meta/Lagom/Tools/CodeGenerators/LibTimeZone/GenerateTimeZoneData.cpp#L815
https://github.com/SerenityOS/serenity/blob/4d68dcfca0487b7cc0c421932f4d4a1c8c44ee5b/Meta/Lagom/Tools/CodeGenerators/LibLocale/GeneratePluralRulesData.cpp#L669
https://github.com/SerenityOS/serenity/blob/4d68dcfca0487b7cc0c421932f4d4a1c8c44ee5b/Meta/Lagom/Tools/CodeGenerators/LibUnicode/GenerateUnicodeData.cpp#L1510
https://github.com/SerenityOS/serenity/blob/4d68dcfca0487b7cc0c421932f4d4a1c8c44ee5b/Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/main.cpp#L32
...

**Comment by kleinesfilmroellchen** [comment]
The CI build failure for Wasm is something I noticed locally as well; I think that requires a cache bust.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    if (use_object_constructor == UseObjectConstructor::Yes)
+        TRY(append(generator, "@object_name@ = TRY(@class_name@::try_create());"));
+    else
+        TRY(append(generator, "@object_name@ = TRY(try_make_ref_counted<@class_name@>());"));
+
+    // Properties
+    TRY(gml_object.try_for_each_property([&](StringView key, NonnullRefPtr<GUI::GML::JsonValueNode> value) -> ErrorOr<void> {
+        auto value_code = TRY(generate_initializer_for(value));
+        auto const& ui_dimension_properties = get_ui_dimension_properties();
+        if (ui_dimension_properties.contains(key)) {
```
Comment (alimpfard) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:224]: This would be nicer as
```suggestion
        if (is_ui_dimension_property(key)) {
```

where you rewrite your `get_ui_dimension_properties` function as
```c++
static bool is_ui_dimension_property(StringView key) {
    HashMap... properties;
    if (...) { ... }
    return properties.contains(key);
}
```

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+#include <LibGUI/GML/Parser.h>
+#include <LibMain/Main.h>
+
+enum class UseObjectConstructor : bool {
+    No,
+    Yes,
+};
+
+HashMap<StringView, StringView> const& get_class_file_mappings();
+HashMap<StringView, StringView> const& get_property_to_type_mappings();
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:24]: All the strings are static Stringviews and they should be used as-is without copying allocation

File: Base/usr/share/man/man5/GML/Usage.md [hunk 7]
```diff
@@ -6,29 +6,31 @@ GML Usage
...
 
 ## C++
 
-You can then reference the variable you set (e.g. `my_app_gml`) in your main C++ file using `load_from_gml()`.
+The C++ source file that is generated will provide an implementation for the `ErrorOr<NonnullRefPtr<App::Widget>> App::widget::try_create()` function, given that the root widget of the GML file is `App::Widget`. For this to work, you have to add a declaration for this function in the header of the `App::Widget` class. (The function will not collide with functions provided by `Core::Object`, do not worry.) Additionally, there must be a public no-argument constructor in `App::Widget`, which is used by the `try_create` implementation.
 
-```cpp
-load_from_gml(my_app_gml);
-```
+Calling the `try_create` function directly or indirectly (e.g. `Window::try_set_main_widget`) will now automatically load the structure defined in GML. Any further initialization should be done in a separate initializer function which you call manually.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Base/usr/share/man/man5/GML/Usage.md:33]: It's required for many things such as action callbacks, models, etc.

**Comment by kleinesfilmroellchen** [comment]
Some more static verification: UI dimensions are now checked and created at compile time as well.

**Review by me-minus — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGUI/Margins.h
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibGUI/Margins.h:1]: Okay, if I unpacked that correctly, you want a comment on the auto-generated method. We don't use doxygen or anything like that.

[the remark this answers is not in the document]
Comment (me-minus) [reply Userland/Libraries/LibGUI/Margins.h:1]: Parse Error ;-)

I want each new  constructor in Margins.h have  the comment  `// GML compatibility constructors, only for use in auto-generated code.' 

The comment at the top of the block is easy to miss

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+// FIXME: Since normal string-based properties take either String or StringView (and the latter can be implicitly constructed from the former),
+//        we need to special-case DeprecatedString property setters while those still exist.
+//        Please remove a setter from this list once it uses StringView or String.
+static bool takes_deprecated_string(StringView property)
+{
+    static HashTable<StringView> deprecated_string_properties;
+    if (deprecated_string_properties.is_empty()) {
+        deprecated_string_properties.set("name"sv);
+    }
+    return deprecated_string_properties.contains(property);
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:76]: It is indeed expected to grow as we transition more to String and these special cases become necessary.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        class_file_mappings.set("GUI::VerticalProgressbar"sv, "GUI/Progressbar"sv);
+    }
+    return class_file_mappings.get(clazz);
+}
+
+// Properties which don't take a direct JSON-like primitive (StringView, int, bool, Array etc) as arguments and need the arguments to be wrapped in a constructor call.
+static Optional<StringView> map_property_to_type(StringView property)
+{
+    static HashMap<StringView, StringView> property_to_type_mappings;
+    if (property_to_type_mappings.is_empty()) {
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:45]: This is even more permanent than the Deprecatedstring one

File: Userland/Applications/Help/MainWidget.h [hunk 10]
```diff
@@ -19,12 +19,13 @@ class MainWidget final : public GUI::Widget {
 public:
     virtual ~MainWidget() override = default;
 
+    MainWidget();
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Help/MainWidget.h:22]: Yes, because we use `try_make_ref_counted` for construction.

File: Userland/Libraries/LibGUI/Margins.h [hunk 11]
```diff
@@ -42,6 +42,29 @@ class Margins {
...
+
+    explicit Margins(Array<i64, 2> vertical_horizontal)
+        : Margins(vertical_horizontal[0], vertical_horizontal[1])
+    {
+    }
+
+    explicit Margins(Array<i64, 3> top_horizontal_bottom)
+        : Margins(top_horizontal_bottom[0], top_horizontal_bottom[1], top_horizontal_bottom[2])
+    {
+    }
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibGUI/Margins.h:61]: Yes, since there's code that currently uses it, and CSS margins shorthands work the same way.

**Comment by kleinesfilmroellchen** [comment]
@BenWiederhake I hope you're happy with my reasoning; some of the weird code you're pointing out is compatibility that can be removed easily.

File: Userland/Libraries/LibGUI/Margins.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply Userland/Libraries/LibGUI/Margins.h:61]: Oof! I was hoping it was not needed anymore. Thanks for double-checking.

File: Userland/Applications/Help/MainWidget.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (BenWiederhake) [reply Userland/Applications/Help/MainWidget.h:22]: It uses `C_OBJECT` and provides a `try_create`, and yet has to have a public constructor. Why are both necessary? Surely there's only one place that creates the Help(1)-specific main widget, or are there multiple?

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        generator.appendln(text);
+        return {};
+    };
+
+    generator.append(TRY(String::repeated(' ', (indentation - 1) * 4)).bytes_as_string_view());
+    generator.appendln("{");
+    if (use_object_constructor == UseObjectConstructor::Yes)
+        TRY(append(generator, "@object_name@ = TRY(@class_name@::try_create());"));
+    else
+        TRY(append(generator, "@object_name@ = TRY(try_make_ref_counted<@class_name@>());"));
```
Comment (BenWiederhake) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:200]:
```suggestion
        TRY(append(generator, "@object_name@ = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) @class_name@ ()));"));
```

This would allow us to make the constructors private, preventing weird refcounting errors.

That's not even hypothetical: We used to have these constructors public by default, and a lot of programs constructed refcounted widgets on the stack, causing all kinds of weird things. (Mostly because something took an additional reference, dropped it, the widget saw that its refcount was "again" zero, thought that it should delete itself, called `free()` on stack memory, boom!)

Let's keep those widget constructors private.

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
One line is all that it takes to keep those widget constructors private. Having them public caused problems in the past, let's not re-introduce that class of bugs.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp (hunk 12 above)
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:200]: I am starting to think we should get rid of `make_ref_counted`, since that always requires a public constructor on a refcounted class...

Comment (BenWiederhake) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:200]: That's a good point.

Hmm, maybe a macro instead? The `adopt_nonnull_{ref,own}_or_enomem` incantation appears in surprisingly many places. Either way, let's avoid public constructors here, and a different PR could solve that problem. If you're not feeling like it, tell me, then I'll give it a stab :D

Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:200]: I'll use your suggestion for now though, you're right.

**Review by BenWiederhake — APPROVED** [review]
Hooray, private-remaining constructors! :partying_face: 

LGTM!

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+    TRY(gml_object.try_for_each_child_object([&](auto const& child) -> ErrorOr<void> {
+        auto child_generator = generator.fork();
+        auto child_variable_name = TRY(next_child_name());
+        child_generator.set("child_variable_name", child_variable_name.bytes_as_string_view());
+        child_generator.set("child_class_name", child.name());
+        TRY(append(child_generator, "RefPtr<@child_class_name@> @child_variable_name@;"));
+        TRY(generate_loader_for_object(child, child_generator.fork(), child_variable_name, indentation + 1, UseObjectConstructor::Yes));
+
+        // Handle the current two special cases of child adding.
+        if (gml_object.name() == "GUI::ScrollableContainerWidget")
```
Comment (timschumi) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:247]: Comparison value can be a StringView literal.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+            TRY(append(child_generator, "static_ptr_cast<GUI::ScrollableContainerWidget>(@object_name@)->set_widget(*@child_variable_name@);"));
+        else if (gml_object.name() == "GUI::TabWidget")
```
Comment (timschumi) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:249]: Same here.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    };
+    TRY(gml_object.try_for_each_child_object([&](auto const& child) -> ErrorOr<void> {
+        auto child_generator = generator.fork();
+        auto child_variable_name = TRY(next_child_name());
+        child_generator.set("child_variable_name", child_variable_name.bytes_as_string_view());
+        child_generator.set("child_class_name", child.name());
+        TRY(append(child_generator, "RefPtr<@child_class_name@> @child_variable_name@;"));
+        TRY(generate_loader_for_object(child, child_generator.fork(), child_variable_name, indentation + 1, UseObjectConstructor::Yes));
+
+        // Handle the current two special cases of child adding.
```
Comment (timschumi) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:246]: Should this have a FIXME?

File: Meta/CMake/code_generators.cmake (hunk 1 above)
Comment (timschumi) [inline Meta/CMake/code_generators.cmake:5]: Maybe do the `compile_gml` -> `stringify_gml` change in a separate commit, makes it less jarring to look at the scroll bar while reviewing. :^)

File: Userland/Applications/Help/CMakeLists.txt [hunk 16]
```diff
@@ -5,17 +5,14 @@ serenity_component(
...
 set(SOURCES
     History.cpp
     main.cpp
     MainWidget.cpp
     ManualModel.cpp
-)
-
-set(GENERATED_SOURCES
-    HelpWindowGML.h
+    HelpWindowGML.cpp
```
Comment (timschumi) [inline Userland/Applications/Help/CMakeLists.txt:15]: Shouldn't this still be in `GENERATED_SOURCES`?

File: Base/usr/share/man/man5/GML/Usage.md [hunk 17]
```diff
@@ -6,43 +6,47 @@ GML Usage
 
 How to use GML in SerenityOS C++ applications
 
+## Note
+
+This manpage describes the new method of loading GML files via generated C++ code. There also is the runtime `load_from_gml` function which has the same behavior. This method should not be used except for specific use cases such as [GMLPlayground](help://man/1/Applications/GMLPlayground).
```
Comment (timschumi) [inline Base/usr/share/man/man5/GML/Usage.md:11]: Is this just me being ultra-pedantic (or wrong) with my English skills, or is the "this" in the last sentence either ambiguous or referring to the wrong loading method entirely?

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+#include <LibGUI/GML/Parser.h>
+#include <LibGUI/UIDimensions.h>
+#include <LibMain/Main.h>
+
+enum class UseObjectConstructor : bool {
+    No,
+    Yes,
+};
+
+static ErrorOr<void> generate_loader_for_object(GUI::GML::Object const& gml_object, SourceGenerator generator, String object_name, size_t indentation, UseObjectConstructor use_object_constructor);
```
Comment (timschumi) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:24]: Is this declaration needed? It seems like the next usage of `generate_loader_for_object` will be the definition itself either way.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+ */
+
+)~~~";
+
+static char const function_start[] = R"~~~(
+// Creates a @main_class_name@ and initializes it.
+// This function was auto-generated by the GML compiler.
+ErrorOr<NonnullRefPtr<@main_class_name@>> @main_class_name@::try_create()
+{
+    RefPtr<@main_class_name@> main_class;
```
Comment (timschumi) [inline Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:120]: Do we have a better fitting name for this, even if it's just `main_object`?

File: Base/usr/share/man/man5/GML/Usage.md [hunk 20]
```diff
@@ -6,43 +6,47 @@ GML Usage
...
 set(SOURCES
-    MyAppGML.h
+    MyAppGML.cpp
 )
 ```
 
 ## C++
 
-You can then reference the variable you set (e.g. `my_app_gml`) in your main C++ file using `load_from_gml()`.
+The C++ source file that is generated will provide an implementation for the `ErrorOr<NonnullRefPtr<App::Widget>> App::Widget::try_create()` function, given that the root widget of the GML file is `App::Widget`. For this to work, you have to add a declaration for this function in the header of the `App::Widget` class. (The function will not collide with functions provided by `Core::Object`, do not worry.) Additionally, there must be a public no-argument constructor in `App::Widget`, which is used by the `try_create` implementation.
```
Comment (timschumi) [inline Base/usr/share/man/man5/GML/Usage.md:31]: The part about the public no-argument constructor doesn't seem accurate anymore. The generated `try_create` method is in the correct namespace to use private constructors, and the Help widget that you converted even makes use of that.

File: Meta/CMake/code_generators.cmake [hunk 21]
```diff
@@ -18,6 +18,22 @@ function(compile_gml source output string_name)
     add_dependencies(all_generated generate_${output_name})
 endfunction()
 
+function(compile_gml source output)
+    set(source ${CMAKE_CURRENT_SOURCE_DIR}/${source})
+    add_custom_command(
+        OUTPUT ${output}
+        COMMAND $<TARGET_FILE:Lagom::GMLCompiler> ${source} > ${output}.tmp
+        COMMAND "${CMAKE_COMMAND}" -E copy_if_different ${output}.tmp ${output}
+        COMMAND "${CMAKE_COMMAND}" -E remove ${output}.tmp
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply Meta/CMake/code_generators.cmake:27]: Not quite, only the BindingsGenerator does. Would be nice :yakbait: to propagate that pattern.

File: Userland/Libraries/LibGUI/Margins.h [hunk 22]
```diff
@@ -42,6 +42,29 @@ class Margins {
...
+
+    explicit Margins(Array<i64, 4> margins)
+        : Margins(margins[0], margins[1], margins[2], margins[3])
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibGUI/Margins.h:64]: That would require more special cases in the generator.

File: Base/usr/share/man/man5/GML/Usage.md (hunk 17 above)
Comment (kleinesfilmroellchen) [reply Base/usr/share/man/man5/GML/Usage.md:11]: No, "This" refers to the runtime method, I don't think there's ambiguity here.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+/*
+ * Auto-generated by the GML compiler
+ */
+
+)~~~";
+
+static char const function_start[] = R"~~~(
+// Creates a @main_class_name@ and initializes it.
+// This function was auto-generated by the GML compiler.
+ErrorOr<NonnullRefPtr<@main_class_name@>> @main_class_name@::try_create()
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:118]: *sigh* This name is required by Widget construction and our own generated code here, this is existing behavior, there already was a comment pointing this out.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp (hunk 18 above)
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:24]: Huh, I thought it wasn't.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+#include <LibMain/Main.h>
+
+enum class UseObjectConstructor : bool {
+    No,
+    Yes,
+};
+
+static ErrorOr<void> generate_loader_for_object(GUI::GML::Object const& gml_object, SourceGenerator generator, String object_name, size_t indentation, UseObjectConstructor use_object_constructor);
+
+// Classes whose header doesn't have the same name as the class.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:26]: Most often, but I've also found a couple more while porting other stuff to this generator. The list will expand quite a bit in the future.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+{
+    static HashTable<StringView> ui_dimension_properties;
+    if (ui_dimension_properties.is_empty()) {
+        ui_dimension_properties.set("min_width"sv);
+        ui_dimension_properties.set("max_width"sv);
+        ui_dimension_properties.set("preferred_width"sv);
+        ui_dimension_properties.set("min_height"sv);
+        ui_dimension_properties.set("max_height"sv);
+        ui_dimension_properties.set("preferred_height"sv);
+    }
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:62]: HashTable doesn't have this constructor.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+        deprecated_string_properties.set("name"sv);
+    }
+    return deprecated_string_properties.contains(property);
+}
+
+static ErrorOr<String> include_path_for(StringView class_name)
+{
+    String pathed_name = TRY(TRY(String::from_utf8(class_name)).replace("::"sv, "/"sv, ReplaceMode::All));
+    if (auto mapping = map_class_to_file(class_name); mapping.has_value())
+        pathed_name = TRY(String::from_utf8(mapping.value()));
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:82]: Sure, that will have the string default-initialized and make the code longer, but whatever.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+    return main_class.release_nonnull();
+}
+)~~~";
+
+static ErrorOr<String> generate_initializer_for(Optional<StringView> property_name, JsonValue value)
+{
+    if (value.is_string()) {
+        if (property_name.has_value() && takes_deprecated_string(*property_name))
+            return String::formatted(R"~~~("{}"sv)~~~", value.as_string());
+        return String::formatted(R"~~~(TRY("{}"_string))~~~", value.as_string());
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:134]: Doesn't really remove any of it; String is still not silently convertible from StringView.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 28]
```diff
@@ -0,0 +1,304 @@
...
+    }
+    // No need to handle the smaller integer types separately.
+    if (value.is_integer<i64>())
+        return String::formatted("static_cast<i64>({})", value.as_integer<i64>());
+    if (value.is_integer<u64>())
+        return String::formatted("static_cast<u64>({})", value.as_integer<u64>());
+    if (value.is_bool())
+        return String::formatted("{}", value.as_bool());
+    if (value.is_double())
+        return String::formatted("static_cast<double>({})", value.as_double());
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:144]: I'd hope so, anything else would be surprising behavior. Nobody actually uses doubles for anything in GML at the moment anyways IIRC.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+        }
+        // FIXME: propagate precise error cause
+        if (!child_type.has_value())
+            return Error::from_string_view("Empty JSON array; cannot deduce type."sv);
+
+        StringBuilder initializer;
+        initializer.appendff("Array<{}, {}> {{ "sv, child_type.release_value(), array.size());
+        for (auto const& child_value : array.values())
+            initializer.appendff("{}, ", TRY(generate_initializer_for({}, child_value)));
+        initializer.append("} "sv);
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:175]: That one is accidental, actually

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp (hunk 15 above)
Comment (kleinesfilmroellchen) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:246]: If you want to make add_child virtual, sure! I don't think so.

File: Userland/Applications/Help/CMakeLists.txt (hunk 16 above)
Comment (kleinesfilmroellchen) [reply Userland/Applications/Help/CMakeLists.txt:15]: No, generated sources is apparently only for generated header. The "copy file if different" handles this just fine.

File: Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (timschumi) [reply Meta/Lagom/Tools/CodeGenerators/GMLCompiler/main.cpp:118]: I have looked through existing feedback before reviewing, wasn't able to find anything that referenced this in particular. Given my relative unfamiliarity with LibGUI, I figured I should just ask outright. No action needed. :^)

File: Meta/Lagom/CMakeLists.txt [hunk 30]
```diff
@@ -358,6 +358,17 @@ install(TARGETS LibTimeZone EXPORT LagomTargets)
 # This is used by the BindingsGenerator so needs to always be built.
 add_serenity_subdirectory(Userland/Libraries/LibIDL)
 
+# LibGUI - only GML
+# This is used by the GML compiler and therefore always needed.
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Meta/Lagom/CMakeLists.txt:374]: Having a (rather specific) `LibGUI_GML` library floating around in general Lagom-land felt weird to me, so I just wanted to ask if that had been considered. Just referencing those files as `${SERENITY_PROJECT_ROOT}/Userland/Libraries/LibGUI/GML/Lexer.cpp` (`Parser.cpp` respectively) from inside the GMLCompiler CMake file seemed like a small enough change to me. :^)

File: Userland/Libraries/LibGUI/GML/AST.h [hunk 31]
```diff
@@ -188,6 +188,19 @@ class Object : public ValueNode {
...
+        for (auto const& child : m_properties) {
+            if (is<KeyValuePair>(child)) {
+                auto const& property = static_cast<KeyValuePair const&>(*child);
+                if (property.key() != "layout" && is<JsonValueNode>(property.value().ptr()))
+                    TRY(callback(property.key(), static_ptr_cast<JsonValueNode>(property.value())));
+            }
+        }
+        return {};
+    }
+
```
Comment (timschumi) [inline Userland/Libraries/LibGUI/GML/AST.h:203]: This looks to have ended up in the wrong commit by accident.

Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/GML/AST.h:203]: Is this in the wrong commit?

**Review by timschumi — CHANGES_REQUESTED** [review]
Two last comments from me, plus responses to the comments that we haven't yet marked as resolved.

:shippie:

**Review by timschumi — APPROVED** [review]
:shippie:
